[FLINK-6022] [avro] Use Avro to serialize Avro in flight and in State

StephanEwen opened the pull request:
## What is the purpose of the change

This changes Avro types to be serialized with a proper Avro serializer. The Avro serializer efficiently handles both Specific Records (generated by Avro) and Avro-reflection-based serialization.

In order to maintain backwards compatibility, Avro type info generates actually a wrapping serializer that falls back to a Pojo (or Kryo) serializer when being reconfigured from an old snapshot.

## Brief change log

  - Adds a proper Avro type serializers
  - Adds a backwards-compatible Avro serializer that falls back to Pojo/Kryo on old snapshots
  - Adds a bunch of test

## Verifying this change

  -  Using Avro specific record types in the program and enjoying nice performant execution ;-)
  - Using Avro for Flink state and getting it serialized via Avro, allowing a schema upgrade of state
  - Running the added unit tests

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no)**
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no)**
  - The serializers: **(yes** / no / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no)**
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

aljoscha commented on flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/typeutils/AvroSerializer.java:
```diff
@@ -153,111 +146,215 @@ public T deserialize(T reuse, DataInputView source) throws IOException {
+			}
+			else if (obj != null && obj.getClass() == AvroSchemaSerializerConfigSnapshot.class) {
+				final AvroSchemaSerializerConfigSnapshot that = (AvroSchemaSerializerConfigSnapshot) obj;
+				return this.schemaString.equals(that.schemaString);
```
Ah forget it, that's probably not necessary here and you're checking for proper schema compatibility in the compatibility-check method.

StephanEwen commented on flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/typeutils/AvroSerializer.java:
```diff
@@ -153,111 +146,215 @@ public T deserialize(T reuse, DataInputView source) throws IOException {
+	}
+
+	private void initializeAvro() {
+		final ClassLoader cl = Thread.currentThread().getContextClassLoader();
```
This is the pattern used by the Kryo serializer for a while rather than `type.getClass().getClassLoader()`, because `type` could in theory be a collection class (application class loader) containing the actual type from another class loader.

Might not be possible to happen for Avro, through, not totally sure...

StephanEwen commented on flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/typeutils/BackwardsCompatibleAvroSerializer.java:
```diff
@@ -0,0 +1,215 @@
+	/**
+	 * Creates a new backwards-compatible Avro Serializer, for the given type.
+	 */
+	public BackwardsCompatibleAvroSerializer(Class<T> type) {
```
The other constructor is a copy constructor, would rather not call it...

StephanEwen commented on flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/typeutils/BackwardsCompatibleAvroSerializer.java:
```diff
@@ -0,0 +1,215 @@
+
+			@SuppressWarnings("unchecked")
+			final TypeSerializer<T> pojoSerializer =
+					(TypeSerializer<T>) typeInfo.createPojoSerializer(new ExecutionConfig());
```
I think it should pick up the info from the snapshot configuration, assuming that the Pojo serializer works correctly.

StephanEwen commented:
Thanks for the review, addressing the comments.

I found the problem with the backwards compatibility: It is the Avro version upgrade.
Avro types generated with Avro 1.7.7 (Flink 1.3) and serialized as PoJos don't work when read into PoJos generated from the same Schema in Avro 1.8.2 (Flink 1.4). The Avro class structure changed, breaking the Pojo compatibility.

Avro is binary format compatible, but not class compatible across versions, it seems. Makes sense, for a serialization format.

Now, this brings up a few issues:
  - Users that use generated Avro types from earlier Avro versions would be forced to upgrade (re-create) these classes with the new Avro version, if they upgrade Flink and it pulls in a newer Avro version. We might solve that by removing avro from `flink-dist` and pushing it into "user code space", where users can dependency manage the Avro version they want and need.

  - In light of that, actually retaining Avro/Pojo backwards compatibility seems super complicated. I would almost suggest to not do that, because it seems it will not be transparent for users anyways (with Avro itself breaking compatibility).

StephanEwen commented:
I updated this PR with the following proposed solution:

  1. Avro is always part of the user code space, and hence will be loaded into the user code classloader. This solves multiple problems:
     - It allows users to use a different Avro version compared to the version in a spoiled classpath (for example when Hadoop depenencies pull in another Avro version).
     - It means that Avro class and schema caching is done per user code classloader, not JVM wide. That prevents "X cannot be cast to X" exceptions.

  2. By default, a new pure Avro serializer is used for Avro types (in flight and in state). Since Avro class format changed, having compatibility activated by default is not possible anyways.

  3. If a user wants to recover an old savepoint that happend to have Avro types serialized as PoJos, they need to do the following;
     - Set the Avro version to the same version that they had when creating the savepoint. Since Avro now lives in user code space, that is possible without changing/rebuilding Flink.
     - Pass a Flag to the `AvroTypeInfo` to use a backwards compatible Serializer: `new AvroTypeInfo(myType, true)`.

aljoscha commented:
I would suggest adding that last paragraph in the `release notes` field of FLINK-6022 so we don't forget about this.

aljoscha commented:
👍 

Btw, now users have to manually depend on `flink-avro` in their projects where before they didn't necessarily have to, right? If yes, we should also put it in the release notes.

shashank734 commented:
@StephanEwen Are these changes part of 1.5 or 1.4, Do you have any example how I can use this with states and CEP? Please give me some hint. I have seen test cases of Input and Output only. State evolution is the main issue for us nowadays.

zentol commented:
@shashank734 The commits are contained in 1.4 already. Have you read [this](https://github.com/apache/flink/pull/4943#issuecomment-342156083) comment?

shashank734 commented:
@zentol Thanks, I think it's the wrong place to ask But Actually We have tried to use AvroTypeInfo, But it was unable to restore from the savepoint (Note we have changed the schema and class with 1 extra variable) So why I was asking if I can get a very minimal example or hint to check Am I am doing something wrong? I am using Scala.

shashank734 commented:
I think this is due to this issue :  https://issues.apache.org/jira/browse/FLINK-9202
